parity/run.rs

```diff
@@ -687,29 +682,13 @@ fn execute_impl<Cr, Rr>(cmd: RunCmd, logger: Arc<RotatingLogger>, on_client_rq:
 		chain_notify.start();
 	}
 
-	let contract_client = {
-		struct FullRegistrar { client: Arc<Client> }
-		impl RegistrarClient for FullRegistrar {
-			type Call = Asynchronous;
-			fn registrar_address(&self) -> Result<Address, String> {
-				self.client.registrar_address()
-					.ok_or_else(|| "Registrar not defined.".into())
-			}
-			fn call_contract(&self, address: Address, data: Bytes) -> Self::Call {
-				Box::new(self.client.call_contract(BlockId::Latest, address, data).into_future())
-			}
-		}
-
-		Arc::new(FullRegistrar { client: client.clone() })
-	};
-
 	// the updater service
 	let updater_fetch = fetch.clone();
 	let updater = Updater::new(
 		&Arc::downgrade(&(service.client() as Arc<dyn BlockChainClient>)),
 		&Arc::downgrade(&sync_provider),
 		update_policy,
-		hash_fetch::Client::with_fetch(contract_client.clone(), updater_fetch, runtime.executor())
+		hash_fetch::Client::with_fetch(client.clone(), updater_fetch, runtime.executor())
```

Comment:
I wonder if this should be a `Weak<dyn RegistrarClient>` so we don't hold on to a hard ref to the `Client`?

Reply:
Good catch, it should most likely be a `Weak` reference given that other stuff is passed as `Weak` too.

Reply:
It looks like the `BlockChainClient` and `SyncProvider` references handle lack of backing client without emitting an error (using a default value, or just not performing the operation).

It would be easy enough to let the `hash_fetch::Client` fail if the backing client isn't there, but this would then be indistinguishable from some other error occurring.

Should we distinguish between the client disappearing, and other errors with the fetcher? (And presumably take no action if the client isn't present.)

Reply:
The latest commit `164c61d` uses a weak reference to the client and fails when it's not present (so doesn't distinguish).

Reply:
I think you're saying that if [this](https://github.com/tobz1000/parity-ethereum/blob/2f0ef2e2dbc4458856c4f57b1ae7b6cf95aaf372/updater/hash-fetch/src/urlhint.rs#L164) fails the the upgrade would fail? I think what you did is fine but I admit I'm not familiar with the code.
/cc @tomusdrw

Reply:
Yes - compare that to [this](https://github.com/tobz1000/parity-ethereum/blob/2f0ef2e2dbc4458856c4f57b1ae7b6cf95aaf372/updater/src/updater.rs#L689) (performs no action if Weak upgrade fails) and [this](https://github.com/tobz1000/parity-ethereum/blob/2f0ef2e2dbc4458856c4f57b1ae7b6cf95aaf372/updater/src/updater.rs#L511) (ultimately performs no action). Neither of these will emit any error if the client is missing.

Reply:
Thank you for explaining. It’s obviously not ideal to swallow errors like this but the changes here do not make things worse so let’s merge.